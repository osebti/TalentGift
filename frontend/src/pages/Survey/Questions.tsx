import Button from "@/components/Button";
import { Answer, Question } from "@/Types";
import { memo } from "react";
import { twMerge } from "tailwind-merge";

interface SurveyQuestionProps {
  question: Question;
  showPrevious: boolean;
  showSubmit: boolean;
  selectedAnswer: Answer | null;
  setAnswer: (answer: Answer) => void;
  nextQuestion: () => void;
  previousQuestion: () => void;
  submitAnswers: () => void;
}

interface AnswerListProps {
  answers: Answer[];
  selectedAnswer: Answer | null;
  setAnswer: (answer: Answer) => void;
}

interface ButtonSectionProps {
  disableNext: boolean;
  showPrevious: boolean;
  showSubmit: boolean;
  selectedAnswer: Answer | null;
  numAnswers: number; // number of choices for a question
  nextQuestion: () => void;
  previousQuestion: () => void;
  submitAnswers: () => void;
}

function AnswerList(props: AnswerListProps) {
  const { answers, selectedAnswer, setAnswer } = props;

  return answers ? (
    <ul
      id="answerContainer"
      className="w-full flex flex-col justify-center items-center gap-4"
    >
      {answers.map((elem, index) => (
        <li className="w-[60%]" key={"answer-" + index}>
          <Button
            label={elem.text}
            id={`answer-${index}${elem == selectedAnswer ? "-selected" : ""}`}
            onClick={() => {
              setAnswer(elem);
            }}
            capitalize={false}
            style={twMerge(
              "w-full min-h-[4rem] rounded-xl border-0",
              index % 2 === 0
                ? elem == selectedAnswer
                  ? "btn-primary text-white"
                  : "bg-primaryTransparent text-black hover:bg-darkPurple hover:text-white"
                : elem == selectedAnswer
                ? "btn-accent text-white"
                : "bg-accentTransparent text-black hover:bg-green hover:text-white"
            )}
          />
        </li>
      ))}
    </ul>
  ) : (
    <></>
  );
}

function ButtonSection(props: ButtonSectionProps) {
  const {
    numAnswers,
    selectedAnswer,
    nextQuestion,
    previousQuestion,
    showPrevious,
    showSubmit,
    disableNext,
    submitAnswers,
  } = props;
  return (
    <div className="flex justify-center items-center gap-4 mb-6">
      {/* Back Button */}
      {showPrevious && (
        <Button
          label="previous"
          id="previous"
          onClick={() => {
            previousQuestion();
          }}
          style="btn w-[8rem] btn-secondary px-12 hover:bg-darkPurple transition"
        />
      )}
      {/* Next/Submit Button */}
      <Button
        label={showSubmit ? "submit" : "next"}
        id={showSubmit ? "submit" : "next"}
        onClick={
          showSubmit
            ? () => {
                submitAnswers();
              }
            : () => {
                selectedAnswer?.index !== -1 || numAnswers === 0
                  ? nextQuestion()
                  : () => {};
              }
        }
        style={twMerge(
          "btn w-[8rem] bg-green px-12 hover:bg-darkGreen transition",
          disableNext ? "btn-disabled" : ""
        )}
      />
    </div>
  );
}

export const SurveyQuestionMemo = memo(function SurveyQuestion(
  props: SurveyQuestionProps
) {
  const {
    question,
    showPrevious,
    showSubmit,
    selectedAnswer,
    setAnswer,
    nextQuestion,
    previousQuestion,
    submitAnswers,
  } = props;

  const disableNext =
    selectedAnswer == null && question.answers.length != 0 ? true : false;

  return (
    <div className="w-full flex flex-col justify-center items-center gap-12 px-2">
      <AnswerList
        answers={question.answers}
        setAnswer={setAnswer}
        selectedAnswer={selectedAnswer}
      />
      {/* Buttons */}
      <ButtonSection
        disableNext={disableNext}
        showPrevious={showPrevious}
        showSubmit={showSubmit}
        numAnswers={question.answers.length}
        selectedAnswer={selectedAnswer}
        nextQuestion={nextQuestion}
        previousQuestion={previousQuestion}
        submitAnswers={submitAnswers}
      />
    </div>
  );
});
