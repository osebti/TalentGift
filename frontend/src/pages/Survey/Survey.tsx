import questionBank from "@/data/question-bank.json";
import { Reducer, useEffect, useMemo, useReducer } from "react";
import { useNavigate } from "react-router-dom";
import { SurveyQuestionMemo } from "./Questions";
import { Answer } from "@/Types";
import NotificationButton from "@/components/NotificationButton";
import { useGlobalState } from "@/state/global";
import config from "../../../../config.js";



interface SurveyState {
  levelIndex: number;
  questionIndex: number;
  answers: Answer[][];
}

interface QuestionHeaderProps {
  level: string;
  levelIndex: number;
  questionIndex: number;
  question: string;
}


function QuestionHeader(props: QuestionHeaderProps) {
  return (
    <div className="w-full flex flex-col gap-6 justify-center items-center">
      <h1
        className="text-2xl lg:text-4xl font-bold text-darkPurple"
        data-testid={`level-header-${props.levelIndex}`}
      >
        {props.level}
      </h1>
      <p
        className="lg:max-w-[50%] max-w-[75%] text-center text-md lg:text-xl"
        data-testid={`question-description-${props.levelIndex}-${props.questionIndex}`}
      >
        {props.question}
      </p>
    </div>
  );
}


export default function Survey(props: {companySurvey?: boolean}) {
  const {companySurvey = false} = props;
  const navigate = useNavigate();
  const [user] = useGlobalState("user");
  const uid = user.uid;
  const [org] = useGlobalState("org")
  const oid = org.oid;
  const manager = user.manager;

  const [state, setState] = useReducer<
    Reducer<SurveyState, Partial<SurveyState>>
  >((state, newState) => ({ ...state, ...newState }), {
    levelIndex: 0,
    questionIndex: 0,
    answers: [[]],
  });

  /* Add appropriate questions depending on whether the survey is a user survey
   * or a company survey */
  const questions = useMemo(() => {
    if (companySurvey) {
      return questionBank.company;
    } else if (manager) {
      return [...questionBank.user, ...questionBank.managers];
    } else {
      return [...questionBank.user, ...questionBank.employees];
    }
  }, [companySurvey, manager]);

  const showPrevious = useMemo(() => {
    if (companySurvey && state.questionIndex != 0) {
      return true
    } else {
      return state.levelIndex != 0 ? true : false
    }
  },[companySurvey, state.levelIndex, state.questionIndex])

  const currentQuestion =
    questions[state.levelIndex].questions[state.questionIndex];

  const showSubmit =
    state.levelIndex == questions.length - 1 &&
    state.questionIndex == questions[state.levelIndex].questions.length - 1
      ? true
      : false;

  useEffect(() => {
    const tempAnswers: Answer[][] = [];
    questions.forEach((level) => {
      const tempAgain: Answer[] = [];
      level.questions.forEach(() => {
        tempAgain.push({ text: "", index: -1 });
      });
      tempAnswers.push(tempAgain);
    });

    setState({ answers: tempAnswers });
  }, [questions]);

  const nextQuestion = () => {
    if (
      state.questionIndex ==
      questions[state.levelIndex].questions.length - 1
    ) {
      setState({ levelIndex: state.levelIndex + 1, questionIndex: 0 });
    } else {
      setState({ questionIndex: state.questionIndex + 1 });
    }
  };

  const previousQuestion = () => {
    if (state.questionIndex == 0) {
      setState({
        levelIndex: state.levelIndex - 1,
        questionIndex: questions[state.levelIndex - 1].questions.length - 1,
      });
    } else {
      setState({ questionIndex: state.questionIndex - 1 });
    }
  };

  const setAnswer = (answer: Answer) => {
    const updatedAnswers = state.answers;
    updatedAnswers[state.levelIndex][state.questionIndex] = answer;
    setState({ answers: updatedAnswers });
  };

  // Formats for database
  const submitAnswers = () => {
    const answerPackage = [];
    console.log(state.answers);
    for (let i = 0; i < state.answers.length; i++) {
      for (let j = 0; j < state.answers[i].length; j++) {
        if (state.answers[i][j] != null) {
          answerPackage.push(state.answers[i][j].index);
        }
      }
    }
    console.log("Answer Package: ", answerPackage);

    fetch(`http://${config.hostname}/surveys/take_survey`, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        Survey_id: companySurvey ? 2: 1, // temporary
        oid: oid,
        Employee_id: uid,
        Answers: answerPackage,
      }),
    }).then((res) => {
      if (res.status === 200) {
        navigate("/success", {
          state: {
            message: "Survey Submitted!",
            returnAddress: companySurvey ? '/dashboard' : `/dashboard/report/${uid}`,
            returnName: "Report",
          },
        });
      }
    });
  };

  return (
    <div className="w-full flex flex-col items-center gap-12 px-2 align-middle pb-8">
      <div className="self-end pr-2">
        {state.levelIndex === 0 && manager && !companySurvey ? <NotificationButton /> : null}
      </div>
      <div className="flex justify-center">
        <QuestionHeader
          level={questions[state.levelIndex].level}
          levelIndex={state.levelIndex}
          questionIndex={state.questionIndex}
          question={currentQuestion.question}
        />
      </div>
      <SurveyQuestionMemo
        question={currentQuestion}
        showPrevious={showPrevious}
        showSubmit={showSubmit}
        selectedAnswer={state.answers[state.levelIndex][state.questionIndex]}
        setAnswer={setAnswer}
        nextQuestion={nextQuestion}
        previousQuestion={previousQuestion}
        submitAnswers={submitAnswers}
      />
      {currentQuestion?.img?.src && (
        <img
          className="mt-12 w-40% sm:max-w-[32rem] md:w-30% lg:w-20% object-contain"
          data-testid={`question-image-${state.levelIndex}-${state.questionIndex}`}
          {...currentQuestion.img}
        />
      )}
    </div>
  );
}
