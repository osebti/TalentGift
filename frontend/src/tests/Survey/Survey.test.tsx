import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";

import Survey from "@/pages/Survey/Survey";
import { BrowserRouter } from "react-router-dom";

afterEach(cleanup);

// Tests for Survey.tsx
// Testing:
// - Survey Rendering (header, question, images)
// - Survey Methods (next, previous, setAnswer)
// - Survey Submit (submit button, submitAnswers)

global.fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
  })
) as jest.Mock;

beforeEach(() => {
  jest.clearAllMocks();
});

describe("Survey Rendering", () => {
  it("renders correctly", () => {
    const page = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    expect(page).toMatchSnapshot();
  });

  it("renders header", () => {
    const { queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    expect(queryByTestId("level-header-0")).not.toBeNull();
  });

  it("renders question", () => {
    const { queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    expect(queryByTestId("question-description-0-0")).not.toBeNull();
  });

  it("renders images", () => {
    const { queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    expect(queryByTestId("question-image-0-0")).not.toBeNull();
  });
});

describe("Survey Methods", () => {
  it("shows next level when next button is clicked", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    expect(queryByTestId("next-button")).not.toBeNull();
    fireEvent.click(getByTestId("next-button"));
    expect(queryByTestId("level-header-1")).not.toBeNull();
  });

  it("shows next question when next button is clicked", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId("next-button"));
    expect(queryByTestId("question-description-1-0")).not.toBeNull();
    fireEvent.click(getByTestId("answer-0-button")); // select first answer
    fireEvent.click(getByTestId("next-button"));
    expect(queryByTestId("question-description-1-1")).not.toBeNull();
  });

  it("shows previous level when previous button is clicked", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId("next-button"));
    expect(queryByTestId("level-header-1")).not.toBeNull();
    fireEvent.click(getByTestId("previous-button"));
    expect(queryByTestId("level-header-0")).not.toBeNull();
  });

  it("shows previous question when next button is clicked", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId("next-button"));
    fireEvent.click(getByTestId("answer-0-button")); // select first answer
    fireEvent.click(getByTestId("next-button"));
    expect(queryByTestId("question-description-1-1")).not.toBeNull();
    fireEvent.click(getByTestId("previous-button"));
    expect(queryByTestId("question-description-1-0")).not.toBeNull();
  });

  it("should setAnswer when answer is selected", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId("next-button"));
    fireEvent.click(getByTestId("answer-0-button"));
    expect(queryByTestId("answer-0-button")).toBeNull();
    expect(queryByTestId("answer-0-selected-button")).not.toBeNull();
  });
});

describe("Survey Submit", () => {
  it("should show submit button when last question is answered", () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );

    while (queryByTestId("next-button")) {
      fireEvent.click(getByTestId("next-button"));
      if (queryByTestId("answer-0-button")) {
        fireEvent.click(getByTestId("answer-0-button"));
      }
    }

    expect(queryByTestId("submit-button")).not.toBeNull();
  });

  it("should submit answers when submit button is clicked", async () => {
    const { getByTestId, queryByTestId } = render(
      <BrowserRouter>
        <Survey />
      </BrowserRouter>
    );

    while (queryByTestId("next-button")) {
      fireEvent.click(getByTestId("next-button"));
      if (queryByTestId("answer-0-button")) {
        fireEvent.click(getByTestId("answer-0-button"));
      }
    }
    expect(queryByTestId("submit-button")).not.toBeNull();
    fireEvent.click(getByTestId("submit-button"));

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(1);
    })
  });
});
