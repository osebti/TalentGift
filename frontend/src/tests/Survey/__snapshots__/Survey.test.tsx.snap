// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey Rendering renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full flex flex-col items-center gap-12 px-2 align-middle pb-8"
      >
        <div
          class="self-end pr-2"
        />
        <div
          class="flex justify-center"
        >
          <div
            class="w-full flex flex-col gap-6 justify-center items-center"
          >
            <h1
              class="text-2xl lg:text-4xl font-bold text-darkPurple"
              data-testid="level-header-0"
            >
              Call to adventure
            </h1>
            <p
              class="lg:max-w-[50%] max-w-[75%] text-center text-md lg:text-xl"
              data-testid="question-description-0-0"
            >
              These are difficult times for some people. The water is dangerous and unpredictable. This ship is ready to sail and the mission needs you! Get ready!
            </p>
          </div>
        </div>
        <div
          class="w-full flex flex-col justify-center items-center gap-12 px-2"
        >
          <ul
            class="w-full flex flex-col justify-center items-center gap-4"
            id="answerContainer"
          />
          <div
            class="flex justify-center items-center gap-4 mb-6"
          >
            <button
              class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn w-[8rem] bg-green px-12 hover:bg-darkGreen transition"
              data-testid="next-button"
              id="next-button"
            >
              next
            </button>
          </div>
        </div>
        <img
          alt="Call To Adventure"
          class="mt-12 w-40% sm:max-w-[32rem] md:w-30% lg:w-20% object-contain"
          data-testid="question-image-0-0"
          src="/ship.svg"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full flex flex-col items-center gap-12 px-2 align-middle pb-8"
    >
      <div
        class="self-end pr-2"
      />
      <div
        class="flex justify-center"
      >
        <div
          class="w-full flex flex-col gap-6 justify-center items-center"
        >
          <h1
            class="text-2xl lg:text-4xl font-bold text-darkPurple"
            data-testid="level-header-0"
          >
            Call to adventure
          </h1>
          <p
            class="lg:max-w-[50%] max-w-[75%] text-center text-md lg:text-xl"
            data-testid="question-description-0-0"
          >
            These are difficult times for some people. The water is dangerous and unpredictable. This ship is ready to sail and the mission needs you! Get ready!
          </p>
        </div>
      </div>
      <div
        class="w-full flex flex-col justify-center items-center gap-12 px-2"
      >
        <ul
          class="w-full flex flex-col justify-center items-center gap-4"
          id="answerContainer"
        />
        <div
          class="flex justify-center items-center gap-4 mb-6"
        >
          <button
            class="btn btn-md rounded-full capitalize font-normal text-white text-sm md:text-base btn w-[8rem] bg-green px-12 hover:bg-darkGreen transition"
            data-testid="next-button"
            id="next-button"
          >
            next
          </button>
        </div>
      </div>
      <img
        alt="Call To Adventure"
        class="mt-12 w-40% sm:max-w-[32rem] md:w-30% lg:w-20% object-contain"
        data-testid="question-image-0-0"
        src="/ship.svg"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
